### PR TITLE
Add custom engine support

### PR DESCRIPTION
**Pull request by JShep1**
I tried to get as minimal changes as possible to get this functioning. 

I switched the default of specifying engines by their plugin name to their shared library file name.  This was primarily because any custom engine is going to be specified by its shared library filename, and the `engines` map previously mapped from plugin name to the `RenderEngine` pointer.  Given the user will likely desire to simply entire their render engine shared library file name, I made this switch.  Specifying `--render-engine ogre` is still supported, but one can now also specify `--render-engine ignition-rendering-ogre` as well.

Another step in getting https://github.com/ignitionrobotics/ign-rendering/issues/100 resolved.

Depends on https://github.com/ignitionrobotics/ign-gazebo/pull/373

### CODE REVIEW

File: src/RenderEngineManager.cc [hunk 1]
```diff
@@ -75,15 +82,18 @@ class ignition::rendering::RenderEngineManagerPrivate
...
-  /// \brief A map of loaded engine plugins to its plugin name
+  /// \brief A map of loaded engine plugins to its plugin name.
   public: std::map<std::string, std::string> enginePlugins;
 
-  /// \brief Plugin loader for managing render engine plugin libraries
+  /// \brief Plugin loader for managing render engine plugin libraries.
   public: ignition::plugin::Loader pluginLoader;
 
+  /// \brief Environment variable which holds paths to look for engine plugins.
+  public: std::string pluginPathEnv = "IGN_GAZEBO_RENDER_ENGINE_PATH";
```
Comment (iche033) [inline src/RenderEngineManager.cc:95]: I think the `IGN_GAZEBO_RENDER_ENGINE_PATH` should not be harded in ign-rendering as that's an ign-gazebo thing. Maybe an API that lets users set the plugin search path? So in ign-gazebo, it reads the `IGN_GAZEBO_RENDER_ENGINE_PATH` env var and calls an ign-rendering to pass this path to the render engine manager

File: src/RenderEngineManager.cc [hunk 2]
```diff
@@ -375,7 +388,10 @@ bool RenderEngineManagerPrivate::LoadEnginePlugin(
   ignition::common::SystemPaths systemPaths;
 
   // Add default install folder.
+  systemPaths.SetPluginPathEnv(std::string(this->pluginPathEnv));
```
Comment (iche033) [inline src/RenderEngineManager.cc:391]: I think this passes the string `IGN_GAZEBO_RENDER_ENGINE_PATH` but not the 'path' pointed to by the env variable.

**Review by iche033 — COMMENTED** [review]
Have a couple comments about the `IGN_GAZEBO_RENDER_ENGINE_PATH` env variable. Other than that, I am able to load the default engines using `ignition-rendeing-` prefix.

**Comment by JShep1** [comment]
Resolved both of your feedback suggestions @iche033 , this branch now depends on the `ign-gazebo` branch here https://github.com/ignitionrobotics/ign-gazebo/pull/373

**Review by iche033 — COMMENTED** [review]
Running `make test` in the build directory shows a few failing tests. Can you take a look? The tests are passing for me locally when run in the latest `ign-rendering3` branch

File: include/ignition/rendering/RenderEngineManager.hh [hunk 3]
```diff
@@ -130,6 +131,10 @@ namespace ignition
       /// \param[in] _index Index of the render-engine to unregister
       public: void UnregisterEngineAt(unsigned int _index);
 
+      /// \brief Set the plugin path from which render engines can be loaded.
+      /// \param[in] _path The name of the environment variable
```
[the remark this answers is not in the document]
Comment (JShep1) [reply include/ignition/rendering/RenderEngineManager.hh]: f76e873e1460a921988c2d4fa3cec273c51485f0

File: include/ignition/rendering/RenderingIface.hh [hunk 4]
```diff
@@ -125,6 +126,11 @@ namespace ignition
     /// \param[in] _index Index of the render-engine to unregister
     IGNITION_RENDERING_VISIBLE
     void unregisterEngine(const unsigned int _index);
+
+    /// \brief Set the plugin path from which render engines can be loaded.
+    /// \param[in] _path The name of the environment variable
```
[the remark this answers is not in the document]
Comment (JShep1) [reply include/ignition/rendering/RenderingIface.hh]: f76e873e1460a921988c2d4fa3cec273c51485f0

File: src/RenderEngineManager.cc [hunk 5]
```diff
@@ -28,19 +28,26 @@
 #include "ignition/rendering/RenderEngineManager.hh"
 #include "ignition/rendering/RenderEnginePlugin.hh"
 
-/// \brief Private implementation of the RenderEngineManager class
+/// \brief Holds information about an engine
+struct EngineInfo
+{
+  std::string name;
+  ignition::rendering::RenderEngine *engine;
```
[the remark this answers is not in the document]
Comment (JShep1) [reply src/RenderEngineManager.cc:39]: f76e873e1460a921988c2d4fa3cec273c51485f0

**Comment by iche033** [comment]
I took a look at the failing tests and found that they are caused by issue with unloading the rendering engines. I made some fixes in pull request #158. Can you take a look?

**Review by iche033 — APPROVED** [review]

**Pull request merged**